ENH: add fsspec support

martindurant opened the pull request:
Supersedes https://github.com/pandas-dev/pandas/pull/33549
closes #33452

- [x] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Commit:
Add remote file io using fsspec.

Commit:
Attempt refactor and clean

TomAugspurger commented on environment.yml:
```diff
@@ -98,7 +98,7 @@ dependencies:
   - pyqt>=5.9.2  # pandas.read_clipboard
   - pytables>=3.4.2  # pandas.read_hdf, DataFrame.to_hdf
-  - s3fs  # pandas.read_csv... when using 's3://...' path
+  - s3fs  # pandas.read_csv... when using 's3://...' path (also brings in fsspec)
```
I think explicitly list fsspec here, if we're import it.

What should we set the minimum supported version to? We'll add this to a few places in the library (`compat._optional`, docs, ...).

TomAugspurger replied:
Can update this, and the minimum versions?

martindurant commented:
I see some parquet changes here, will need a bit of work...

TomAugspurger commented on pandas/io/common.py:
```diff
@@ -158,11 +144,25 @@ def urlopen(*args, **kwargs):
+    try:
+        import fsspec  # noqa: F401
+
+        return isinstance(url, str) and ("::" in url or "://" in url)
```
Should this also check if there's an fsspec-compatible implementation for that protocol? What's the behavior / error message if you have `fsspec` installed but not `s3fs` and do `read_csv("s3://...")`?

martindurant replied:
fsspec gives specific error messages for known protocols that are not available due to missing dependency (s3fs would be such a one), and a different message is the protocol is completely unknown.

TomAugspurger replied:
Does that error surface to the user? Or does pandas swallow it somewhere along the way?

TomAugspurger commented on pandas/io/common.py:
```diff
@@ -158,11 +144,25 @@ def urlopen(*args, **kwargs):
     encoding: Optional[str] = None,
     compression: Optional[str] = None,
     mode: Optional[str] = None,
+    **storage_options,
```
Add a type? `Dict[str, Any]` I think?

TomAugspurger replied:
Does `Dict[str, Any]` work?

TomAugspurger commented on pandas/io/common.py:
```diff
@@ -175,6 +175,7 @@ def get_filepath_or_buffer(
     compression : {{'gzip', 'bz2', 'zip', 'xz', None}}, optional
     encoding : the encoding to use to decode bytes, default is 'utf-8'
     mode : str, optional
+    storage_options: passed on to fsspec, if using it
```
```suggestion
    **storage_options : dict, optional
        passed on to fsspec.open, if using it.
```

jreback replied:
add a versionadded tag (1.1)

TomAugspurger commented on pandas/tests/io/test_fsspec.py:
```diff
@@ -0,0 +1,70 @@
+
+def test_to_csv(cleared_fs):
+    df1.to_csv("memory://test/test.csv", index=True)
+    gc.collect()  # pandas does not explicitly close file buffers
```
(why) is this necessary?

martindurant replied:
I'll double check - maybe this was only need by the previous version of fsspec

TomAugspurger commented on pandas/tests/io/test_fsspec.py:
```diff
@@ -0,0 +1,70 @@
+
+
+@pytest.fixture
+@td.skip_if_installed("fsspec")
```
Why is this skipped if fsspec is installed?

martindurant replied:
Ah, should be the opposite - but why did the tests run for me??

TomAugspurger commented on pandas/tests/io/test_gcs.py:
```diff
@@ -1,115 +0,0 @@
-from io import StringIO
```
I think we still wantthis test file.

martindurant replied:
The GCS and S3 tests were not really testing either of those APIs, they both had (different) FS mocks which were a lot like fsspec's memory FS.

TomAugspurger replied:
Do the tests pass with the new implementation, or are the mocks not valid?

At the least, we'll want replacements for each of these tests.

martindurant replied:
That's what test_fsspec.py  does. The existing tests in the two files were essentially the same with mock FSs (i.e., didn't actually touch the gcsfs or s3fs code at all).
Actually, I notice that moto is listed in several of the test envs, so we should be able to build out a real suite for s3 and maybe a couple of select fsspec protocols which don't have dependencies (file, memory and ... ?).

TomAugspurger commented on pandas/tests/io/test_pickle.py:
```diff
@@ -456,42 +456,10 @@ def mock_urlopen_read(*args, **kwargs):
-                return f
-
-        monkeypatch.setattr("s3fs.S3FileSystem", MockS3FileSystem)
+@td.skip_if_no("fsspec")
```
I think we'll still want the old tests.

jorisvandenbossche replied:
@TomAugspurger is this one still relevant, or can it be "resolved" ?

TomAugspurger commented on pandas/tests/io/test_s3.py:
```diff
@@ -1,25 +0,0 @@
-from io import BytesIO
```
Again: ideally we keep the old tests, aside from the `is_s3_url` one.

TomAugspurger reviewed (commented):
We'll want a whatsnew in 1.1.0.rst. You can add a subsection to enhancements for this.

Commit:
Merge branch 'master' into feature/add-fsspec-support

pep8speaks commented:
Hello @martindurant! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/io/parquet.py`](https://github.com/pandas-dev/pandas/blob/302ba1307c193aeccd1d1787b5725609fabb9d9c/pandas/io/parquet.py):

> [Line 132:22](https://github.com/pandas-dev/pandas/blob/302ba1307c193aeccd1d1787b5725609fabb9d9c/pandas/io/parquet.py#L132): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','

Commit:
readd and adapt s3/gcs tests

Commit:
remove gc from test

Commit:
Simpler is_fsspec

Commit:
add test

martindurant commented:
(NB: some failures are due to v0.7.4 of fsspec only being on conda-forge)

martindurant commented:
but I will need some help with
```
pandas/tests/io/test_fsspec.py:18: error: Item "None" of "Optional[str]" has no attribute "encode"
```
(i.e., I know that, given the input, the output is str and not None, but how do I tell mypy that)

jreback commented on pandas/tests/io/test_fsspec.py:
```diff
@@ -0,0 +1,105 @@
+from pandas.util import _test_decorators as td
+
+from pandas.io.common import is_fsspec_url
+
```
everything in this module is safe if fsspec is not available?

jreback replied:
ok to simply skip the module

martindurant replied:
you mean `fsspec = pytest.importorskip("fsspec")`? 
`test_is_fsspec_url` should be moved to test_common in that case.

jreback commented on pandas/tests/io/test_parquet.py:
```diff
@@ -543,8 +543,6 @@ def test_s3_roundtrip(self, df_compat, s3_resource, pa):
     @td.skip_if_no("s3fs")
     @pytest.mark.parametrize("partition_col", [["A"], []])
```
parquet tests will now need to skip if fsscpec is not installed. 

However we DO seem to depend on it in places, so ok with adding this requirement (pls add in io.rst and the doc-strings). e.g. need either fastparquet / pyarrow AND fsspec.

martindurant replied:
Parquet read/write local and read from http will not be affected

martindurant replied:
(plus, `s3fs` depends on `fsspec`, so those that depend on the former will already be skipped)

TomAugspurger commented on pandas/io/parquet.py:
```diff
@@ -107,6 +102,11 @@ def write(
         # write_to_dataset does not support a file-like object when
         # a directory path is used, so just pass the path string.
         if partition_cols is not None:
+            if is_fsspec_url(path) and "filesystem" not in kwargs:
```
Can you leave a comment explaining this `"filesystem" not in kwargs` check? It's not obvious to me why it's needed.

martindurant replied:
In `fsspec`, you can specify the exact protocol you would like beyond that inferred from the URL. Given that we don't pass storage_options through yet, perhaps this gives more flexibility than required and I can remove it.

martindurant replied:
Sorry, edit on that: this is the filesystem parameter (i.e., an actual instance) to pyarrow. I have no idea if people might currently be using that.

TomAugspurger replied:
Ah, you're saying the user could pass a filesystem like

```
df.to_parquet(..., filesystem=filesystem)
```

That certainly seems possible. Could you ensure that we have a test for that?

TomAugspurger commented on pandas/tests/io/test_gcs.py:
```diff
@@ -73,6 +76,10 @@ def mock_get_filepath_or_buffer(*args, **kwargs):
 def test_to_parquet_gcs_new_file(monkeypatch, tmpdir):
     """Regression test for writing to a not-yet-existent GCS Parquet file."""
+    from fsspec import AbstractFileSystem
+    from fsspec.registry import _registry
```
Can we avoid relying on private attributes?

TomAugspurger commented on pandas/io/parquet.py:
```diff
@@ -185,17 +188,14 @@ def write(
-                parquet_file = self.api.ParquetFile(path, open_with=filesystem.open)
-            finally:
-                s3.close()
+            # if path is s3:// or gs:// we need to open the file in 'wb' mode.
```
Is this comment accurate? Should `wb` be `rb`?

TomAugspurger commented on pandas/io/parquet.py:
```diff
@@ -185,17 +188,14 @@ def write(
-            finally:
-                s3.close()
+            # if path is s3:// or gs:// we need to open the file in 'wb' mode.
+            # TODO: Support 'ab'
```
Bad copy-paste on this comment?

martindurant replied:
This was left over from previous code and no longer relevant (it shows up because of changed indentation).

TomAugspurger reviewed (commented):
Linting error: https://github.com/pandas-dev/pandas/pull/34266/checks?check_run_id=697447968#step:11:10

Can you run mypy on this locally?

CI failures at https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=35892&view=logs&j=98c84d08-34d8-513e-80be-2c581992dd5a&t=f75024c1-aff0-57fb-4f0e-35923b654e09&l=208. Do we need to specify a minimum version of fsspec?

cc @gfyoung if you have a chance to review.

martindurant commented on pandas/io/common.py:
```diff
@@ -213,19 +183,14 @@ def get_filepath_or_buffer(
+        file_obj = fsspec.open(
+            filepath_or_buffer, mode=mode or "rb", **storage_options
+        ).open()
+        # TODO: both fsspec and pandas handle compression and encoding
```
Given that pandas must still handle compression and encoding for local and http, that code will not be deprecated. Therefore, I think it's fine that we don't advertise the fact that fsspec can do that part too, and open everything on the backend as "rb"/"wb", uncompressed. The TODO would be resolved if at some point we decided that fsspec should handle *all* file ops, which is not likely in the near term.

TomAugspurger replied:
Agreed, so I think the TODO can be removed.

Commit:
Answered most points

martindurant commented:
> Linting error: https://github.com/pandas-dev/pandas/pull/34266/checks?check_run_id=697447968#step:11:10

I need help with this. The return of `to_csv` is Optional[str], but for the input I gave it, it must be str, so that encode is a legal thing to do. How do I specify that? Wrapping it in `str()` does not seem right. 

> CI failures

There runs are using an older fsspec from conda defaults. I can ask the conda team to update the version, but we should be certain that this PR doesn't require any further changes in fsspec.

TomAugspurger commented on environment.yml:
```diff
@@ -98,7 +98,7 @@ dependencies:
   - pyqt>=5.9.2  # pandas.read_clipboard
   - pytables>=3.4.3  # pandas.read_hdf, DataFrame.to_hdf
-  - s3fs  # pandas.read_csv... when using 's3://...' path
+  - s3fs>=0.4.0  # pandas.read_csv... when using 's3://...' path (also brings in fsspec)
```
Can you also include fsspec at it's minimum supported version here?

TomAugspurger replied:
Also add to `pandas.compat._optional` and  the table in `doc/source/install.rst`.

TomAugspurger commented on pandas/io/common.py:
```diff
@@ -213,19 +184,14 @@ def get_filepath_or_buffer(
-            filepath_or_buffer, encoding=encoding, compression=compression, mode=mode
-        )
+    if is_fsspec_url(filepath_or_buffer):
+        import fsspec
```
Use `pandas.compat._optional.import_optional_dependency("fsspec")`

TomAugspurger commented on pandas/tests/io/test_fsspec.py:
```diff
@@ -0,0 +1,98 @@
+    import fsspec
+
+    memfs = fsspec.filesystem("memory")
+    try:
+        yield memfs
+    finally:
+        memfs.store.clear()
```
the `try` / `finally` is unnecessary. pytest ensures that the cleanup always runs: https://docs.pytest.org/en/latest/fixture.html#fixture-finalization-executing-teardown-code.

TomAugspurger commented on pandas/tests/io/test_fsspec.py:
```diff
@@ -0,0 +1,98 @@
+        memfs.store.clear()
+
+
+@td.skip_if_no("fsspec")
```
Can you `pytest.importorskip` inside the `cleared_fs` fixture? Then you don't need to repeat this.

TomAugspurger commented on pandas/io/parquet.py:
```diff
@@ -185,17 +188,11 @@ def write(
-        if is_s3_url(path):
-            from pandas.io.s3 import get_file_and_filesystem
+        if is_fsspec_url(path):
+            import fsspec
```
Use `import_optional_dependency`

TomAugspurger commented on pandas/io/parquet.py:
```diff
@@ -107,6 +102,11 @@ def write(
         # a directory path is used, so just pass the path string.
         if partition_cols is not None:
+            if is_fsspec_url(path) and "filesystem" not in kwargs:
+                import fsspec.core
```
`import_optional_dependency`

TomAugspurger commented on pandas/io/parquet.py:
```diff
@@ -122,9 +122,14 @@ def write(
-            path, filesystem=get_fs_for_path(path), **kwargs
-        )
+        if is_fsspec_url(path) and "filesystem" not in kwargs:
+            import fsspec.core
```
`import_optional_dependency`

TomAugspurger commented on pandas/io/parquet.py:
```diff
@@ -164,13 +169,11 @@ def write(
-            # if path is s3:// or gs:// we need to open the file in 'wb' mode.
-            # TODO: Support 'ab'
+        if is_fsspec_url(path):
+            import fsspec
```
`import_optional_dependency`

TomAugspurger commented on pandas/tests/io/test_parquet.py:
```diff
@@ -558,11 +556,7 @@ def test_s3_roundtrip_for_dir(self, df_compat, s3_resource, pa, partition_col):
-            write_kwargs={
-                "partition_cols": partition_col,
-                "compression": None,
-                "filesystem": get_s3_fs(),
```
What happens if filesystem is provided?

alimcmaster1 replied:
Agree + do we need to update any of the docstrings?

martindurant replied:
It works fine, this is tested

TomAugspurger reviewed (commented):
Can you add a new subsection to "Enhancements" to doc/source/whatsnew/v1.1.0.rst`?

Commit:
Implemented suggestions

Added test with filesystem=

Added whatsnew

Changed imports, updated comments

Commit:
lint

TomAugspurger commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -197,6 +197,20 @@ For a full example, see: :ref:`timeseries.adjust-the-start-of-the-bins`.
+support for several other storage implementations such as Azure Datalake and Blob,
+SSH, FTP, dropbox and github. For docs and capabilities, see the `fsspec docs`_.
+
+In the future, we will implement a way to pass parameters to the invoked
+filesystem instances.
```
I think that can be removed.

TomAugspurger commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -197,6 +197,20 @@ For a full example, see: :ref:`timeseries.adjust-the-start-of-the-bins`.
 
 .. _whatsnew_110.enhancements.other:
 
+fsspec now used for filesystem handling
```
Should go above the `.. _whatsnew_110.enhancements.other`. And you might want to add a header for it.

TomAugspurger replied:
I don't think this was fixed. L198 (` .. _whatsnew_110.enhancements.other: `) needs to be below this new section.

TomAugspurger commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -197,6 +197,20 @@ For a full example, see: :ref:`timeseries.adjust-the-start-of-the-bins`.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For reading and writing to filesystems other than local and reading from HTTP(S),
+the optional dependency ``fsspec`` will be used to dispatch operations. This will give unchanged
```
Link to the original issue at the end of the first sentence.

TomAugspurger replied:
Also not fixed yet.

TomAugspurger commented on pandas/compat/_optional.py:
```diff
@@ -8,8 +8,9 @@
+    "fsspec": "0.7.4",
     "fastparquet": "0.3.2",
-    "gcsfs": "0.2.2",
+    "gcsfs": "0.6.0",
```
Will need to update these in the "Increased minimum versions for dependencies" section of the whatsnew.

Commit:
Add versions info

martindurant commented:
Note: fsspec 0.7.4 will be up on conda defaults soon. At that point I'll remove the WIP tag.

TomAugspurger reviewed (commented):
I don't know that we'll need to wait for the fsspec to be updated on defaults. You can update the various `ci/deps/*.yml` files to include `fsspec>=0.7.4` wherever we include s3fs or gcsfs already. That should force it to use conda-forge till defaults is updated.

Can you also run `scripts/generate_pip_deps_from_conda.py`?

Commit:
Update some deps

martindurant commented:
Those two points were fixed on my local branch, not yet pushed.

martindurant commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -195,6 +195,20 @@ If needed you can adjust the bins with the argument ``offset`` (a Timedelta) tha
+support for several other storage implementations such as Azure Datalake and Blob,
+SSH, FTP, dropbox and github. For docs and capabilities, see the `fsspec docs`_.
+
+.. _GH #33452: https://github.com/pandas-dev/pandas/issues/33452
```
Does the "#" break this link?

TomAugspurger replied:
It should be like

    :issue:`33452`

Commit:
issue link syntax

martindurant commented:
Remaining issues: 
- failure on parsing datetime, yes in S3 test (but I don't get how my changes could have caused this)
- the mypy failure on `df.to_csv().encode()`
- some envs still having old gcsfs, but tests only skipping if there is no gcsfs at all. Should both update skipping and envs.

martindurant commented:
Oh, and one more: the use of `"s3n://"` as a protocol. This is not listed as a protocol for S3 in s3fs, but "s3a" is. Is it really needed? A [good description](https://stackoverflow.com/a/33356421/3821154) of what the protocol specifiers mean to hadoop.

Commit:
More specific test versions

TomAugspurger commented:
s3n comes from https://github.com/pandas-dev/pandas/pull/11071. I'm not sure if it's in use, but I'd like to avoid changing it without warning.

martindurant commented:
fsspec does not yet have an entrypoints method for declaring backends, so I'll have to both add s3n as a "known protocol" and also add it to the list of possible protocols in S3FileSystem, and release both.

TomAugspurger commented on pandas/tests/io/test_fsspec.py:
```diff
@@ -0,0 +1,92 @@
+        "dt": date_range("2018-06-18", periods=2),
+    }
+)
+text = df1.to_csv(index=False).encode()
```
cc @simonjayhawkins, mypy is complaining here since to_csv can technically return None, but it's not here since we don't specify a path. Can we `type: ignore` this?

simonjayhawkins replied:
since the return type of to_csv is Optional[str] and the return value is dependent on the type of path_or_buf rather than the value of path_or_buf, an overload doesn't need Literals, so doing this properly shouldn't be too hard. similar done recently in #34207

to keep this PR focussed, type: ignore is the easiest. A comment for later reference may help.

(could also maybe split the assignment so the ignore is only for the .encode() call.)

or overload as a pre-cursor PR.

TomAugspurger replied:
Oh cool, I didn't know that overload handled this case.

So I think a type: ignore for now, and a followup issue to remove it by overloading the signature.

Commit:
Account for alternate S3 protocols, and ignore type error

Note that the S3 protocols (s3a and s3n) could have been implemented
in s3fs and/or fsspec, but they don't actually have any meaning
outside of hadoop implementations.

Commit:
Add comment to mypy ignore insrtuction

Commit:
more mypy

Commit:
more black

simonjayhawkins commented on pandas/io/common.py:
```diff
@@ -150,38 +137,20 @@ def urlopen(*args, **kwargs):
     encoding: Optional[str] = None,
     compression: Optional[str] = None,
     mode: Optional[str] = None,
+    **storage_options: Dict[str, Any],
```
do we want callers to pass a dict or collect additional kwargs here, the docstring implies the former.

martindurant replied:
We are not passing anything at all yet, so I don't mind whether it's kwargs or a dict keyword. I imagine in a user function like read_csv, there would be a `storage_options={...}`, which is the common usage in Dask and Intake.

simonjayhawkins commented on pandas/io/common.py:
```diff
@@ -150,38 +137,20 @@ def urlopen(*args, **kwargs):
     encoding: Optional[str] = None,
     compression: Optional[str] = None,
     mode: Optional[str] = None,
+    **storage_options: Dict[str, Any],
```
so probably better to use storage_options as a keyword here as well, it gives us stricter checking with mypy.

```suggestion
    storage_options: Dict[str, Any],
```

martindurant commented:
OK, that only leaves the localised timestamps. Any suggestions?

Commit:
Make storage_options a dict rather than swallowing kwargs

jorisvandenbossche commented on pandas/tests/io/test_parquet.py:
```diff
@@ -537,14 +537,24 @@ def test_categorical(self, pa):
+            write_kwargs=kw,
+        )
+
+    def test_s3_roundtrip_explicit_fs(self, df_compat, s3_resource, pa):
```
Naming nitpick: I think it's the test above that has the explicit filesystem passed?

TomAugspurger commented:
Are you able to reproduce the failures locally? To get the exact versions used in the env, you can `conda env create --file=ci/deps/azure-36-locale.yaml`.

The locale setting can be a bit trickier. That test is job is I think it here: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=36700&view=logs&j=98c84d08-34d8-513e-80be-2c581992dd5a&t=db7ca50c-72c8-5416-b299-59f9d25bdff8&l=19 to `it_IT.utf8`. On my Mac I can achieve the same effect (I think) by setting LANG

```
$ LANG=it_IT.UTF-8 python -c 'import locale; print(locale.getlocale())'
('it_IT', 'UTF-8')
```

jorisvandenbossche reviewed (commented):
Looks good! Added a few comments

jorisvandenbossche commented:
Additonal note: https://github.com/pandas-dev/pandas/pull/34500 should go in first I think (otherwise that will be hard to backport)

martindurant commented on doc/source/getting_started/install.rst:
```diff
@@ -267,8 +267,9 @@ SQLAlchemy                1.1.4              SQL support for databases other tha
 SciPy                     0.19.0             Miscellaneous statistical functions
 XLsxWriter                0.9.8              Excel writing
 blosc                                        Compression for HDF5
+fsspec                    0.7.4              File operations handling
```
Hm, it handles other things too, just not "local" or "http(s)". That would make for a long comment.

jorisvandenbossche replied:
Then something like "filesystems other than local or http(s)" isn't too long I would say

martindurant commented on pandas/io/common.py:
```diff
@@ -150,38 +137,20 @@ def urlopen(*args, **kwargs):
-    s3fs.S3FileSystem, gcsfs.GCSFileSystem, None
-        Appropriate FileSystem to use. None for local filesystem.
+    Returns true if fsspec is installed and the given URL looks like
+    something fsspec can handle
```
Hence "looks like"; whether it can actually be handled will depend on fsspec and whichever backend being installed - but you'll get the error shortly after, when trying to use the URL.

martindurant commented on pandas/io/parquet.py:
```diff
@@ -107,6 +102,14 @@ def write(
+                import fsspec.core
+
+                fs, path = fsspec.core.url_to_fs(path)
+                kwargs["filesystem"] = fs
```
For the non-partitioned case, we pass a file-like object directly.

martindurant replied:
We could not process the path into a file object and pass the filesystem in both cases, if preferred.

jorisvandenbossche replied:
> For the non-partitioned case, we pass a file-like object directly.

If that's what we do right now, fine to leave it like that in this PR.

Commit:
More requested changes

martindurant commented:
@TomAugspurger , the following passes for me locally
```
LANG=it_IT.UTF-8 pytest pandas/tests/io/test_fsspec.py::test_from_s3_csv
```

TomAugspurger commented:
@martindurant is that in an environment created from the environment file in `ci/deps`? And are you able to step through and see where things go wrong? If not I can take a look in a few days.

I think our setup also does some hacky stuff to set the locale: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=36798&view=logs&j=a3a13ea8-7cf0-5bdb-71bb-6ac8830ae35c&t=2296ffae-5e67-52bf-66a8-6160145a0f80&l=14, but I hoped that wouldn't be needed.

There might be some additional failures. https://travis-ci.org/github/pandas-dev/pandas/jobs/695142967#L1480 indicates that a file handle may not be being closed correctly.

```
 ResourceWarning: unclosed file <_io.TextIOWrapper name=0 mode='r' encoding='UTF-8'>

sys:1: ResourceWarning: unclosed <socket.socket fd=8, family=AddressFamily.AF_INET, type=2050, proto=0, laddr=('0.0.0.0', 0)>

=================================== FAILURES ===================================

________________________ TestPandasContainer.test_to_s3 ________________________

[gw0] linux -- Python 3.6.6 /home/travis/miniconda3/envs/pandas-dev/bin/python

self = <pandas.tests.io.json.test_pandas.TestPandasContainer object at 0x7f796ca6e908>

s3_resource = s3.ServiceResource()

    def test_to_s3(self, s3_resource):

        # GH 28375

        mock_bucket_name, target_file = "pandas-test", "test.json"

        df = DataFrame({"x": [1, 2, 3], "y": [2, 4, 6]})

        df.to_json(f"s3://{mock_bucket_name}/{target_file}")

>       assert target_file in (

            obj.key for obj in s3_resource.Bucket("pandas-test").objects.all()

        )

E       AssertionError: assert 'test.json' in <generator object TestPandasContainer.test_to_s3.<locals>.<genexpr> at 0x7f796c5c6518>

pandas/tests/io/json/test_pandas.py:1672: AssertionError
```

TomAugspurger commented:
FYI, @martindurant I am only able to reproduce setting in the init

```diff
diff --git a/pandas/__init__.py b/pandas/__init__.py
index d6584bf4f1..1a4147f49d 100644
--- a/pandas/__init__.py
+++ b/pandas/__init__.py
@@ -1,4 +1,6 @@
 # flake8: noqa
+import locale
+locale.setlocale(locale.LC_ALL, 'zh_CN.UTF-8')
 
 __docformat__ = "restructuredtext"
```

In particular, I'm not able to reproduce by setting the LC_ALL, etc. variables

```
$ LC_ALL=zh_CN.UTF-8 LANG=zh_CN.UTF-8 pytest pandas/tests/io/test_fsspec.py::test_from_s3_csv -x --pdb
```

And it's boto that's failing to parse the timestamp. I'm inclined to think that this is an artifact of our test setup. We shouldn't be setting that in the init, since it isn't indicative of real-world usage, but I'd like to better understand why that value is causing issues for boto first.

TomAugspurger commented:
Well, after [some reading](https://docs.python.org/3/library/locale.html#background-details-hints-tips-and-caveats), I'm inclined to say that our tests are "wrong". Python apparently always uses the C locale, even when `LANG`, etc. are set.

I can also reproduce the error on master with a simple test like

```python
import pandas as pd
import locale
locale.setlocale(locale.LC_ALL, 'zh_CN.UTF-8')


def test_foo(s3_resource):
    assert s3_resource
    pd.read_csv('s3://pandas-test/tips.csv')
```

So for now @martindurant I'd recommend removing fsspec / s3fs from CI jobs that set the locale like that. We can investigate followups, though we'll want to make sure we aren't losing coverage on tests.

Still haven't looked at the resource warning failure.

jorisvandenbossche commented on pandas/io/parquet.py:
```diff
@@ -106,7 +100,22 @@ def write(
         table = self.api.Table.from_pandas(df, **from_pandas_kwargs)
         # write_to_dataset does not support a file-like object when
         # a directory path is used, so just pass the path string.
+        if is_fsspec_url(path) and "filesystem" not in kwargs:
+            # TODO: use_legacy_dataset=False will require work, when it is available
+            assert kwargs.get("use_legacy_dataset", False) is False
```
```suggestion
        if is_fsspec_url(path) and "filesystem" not in kwargs and kwargs.get("use_legacy_dataset", True):
            # TODO: use_legacy_dataset=False will require work, when it is available
            assert kwargs.get("use_legacy_dataset", False) is False
```

`use_legacy_dataset` should be True for this code branch, as otherwise it cannot (yet) handle fsspec filesystems

jorisvandenbossche replied:
Ah, actually, it can be removed here, as this is the "write" path. But should be added below in the "read" part (where I put the original comment).

Commit:
Remove fsspec from locale tests

And move "legacy_datasets" clause

Commit:
tweak

Commit:
Merge branch 'master' into feature/add-fsspec-support

Commit:
Merge branch 'master' into feature/add-fsspec-support

jreback commented on ci/deps/azure-36-locale.yaml:
```diff
@@ -15,7 +15,6 @@ dependencies:
 
   # pandas dependencies
   - beautifulsoup4
-  - gcsfs
```
can you re-add fsscspec here as well as leaving these (pin the versions if needed)

martindurant replied:
Sorry, what do you mean? It seems that the current version of boto/moto or something else is causing faulty locale-specific timestamp handling, that's why these were removed (because they are tested in the other runs)

jreback commented on ci/deps/travis-36-locale.yaml:
```diff
@@ -33,7 +32,6 @@ dependencies:
   - pytables
   - python-dateutil
   - pytz
-  - s3fs=0.3.0
```
same comments as above

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -246,6 +246,18 @@ If needed you can adjust the bins with the argument ``offset`` (a Timedelta) tha
 
+fsspec now used for filesystem handling
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
```
this should be in the api breaking section

jorisvandenbossche replied:
I don't think there are actually breaking changes? For s3fs and gcfs, it stays the same (@martindurant is that correct?), and in enables new filesystems as an enhancement?

martindurant replied:
That's right, for gcs:// and s3://, you get the same file system instances you would have.

jreback replied:
oh ok (maybe edit to make that clear then)

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -246,6 +246,18 @@ If needed you can adjust the bins with the argument ``offset`` (a Timedelta) tha
+the optional dependency ``fsspec`` will be used to dispatch operations (:issue:`33452`).
+This will give unchanged
+functionality for S3 and GCS storage, which were already supported, but also add
+support for several other storage implementations such as Azure Datalake and Blob,
```
can you provide a link to azure blob / lake

jreback commented on environment.yml:
```diff
@@ -98,7 +98,8 @@ dependencies:
   - pyqt>=5.9.2  # pandas.read_clipboard
   - pytables>=3.4.3  # pandas.read_hdf, DataFrame.to_hdf
-  - s3fs  # pandas.read_csv... when using 's3://...' path
+  - s3fs>=0.4.0  # pandas.read_csv... when using 's3://...' path
```
gcsfs?

martindurant replied:
Was not included here previously. There are many FS backends we *could* include, in theory, but I don't imagine anyone wants to write tests for them right now.

jreback replied:
sure but we are testing it already, its consistent to add it here

jreback commented on pandas/io/parquet.py:
```diff
@@ -106,7 +100,21 @@ def write(
         # write_to_dataset does not support a file-like object when
         # a directory path is used, so just pass the path string.
+        if is_fsspec_url(path) and "filesystem" not in kwargs:
+            # TODO: use_legacy_dataset=False will require work, when it is available
```
can you create an issue & reference it here

martindurant replied:
Decided to remove it, as the pyarrow side is still being developed, and may well be able to use fsspec implementations after all.

jreback commented on pandas/io/parquet.py:
```diff
@@ -115,16 +123,23 @@ def write(
+            assert kwargs.get("use_legacy_dataset", True) is True
+            import fsspec.core
+
+            fs, path = fsspec.core.url_to_fs(path)
```
does fsspec handle closing itself?

martindurant replied:
The file system does not need to be closed, and files are closed when they fall out of scope

jreback replied:
ok

jreback reviewed (changes requested):
looks good a couple of questions / comments

Commit:
requested changes

Commit:
add gcsfs to environment.yml

TomAugspurger commented:
Can you re-run `scripts/generate_pip_deps_from_conda.py` and commit the output?

Maybe merge master to. travis was failing.

Commit:
rerun deps script

martindurant commented:
@TomAugspurger done

jreback reviewed (approved):
lgtm. @TomAugspurger merge when ready.

jorisvandenbossche commented:
As mentioned before, I would prefer to merge https://github.com/pandas-dev/pandas/pull/34500 first (to ensure this case is actually fixed/tested, and also working with those changes)

jreback commented:
> As mentioned before, I would prefer to merge #34500 first (to ensure this case is actually fixed/tested, and also working with those changes)

sure let's do that

jorisvandenbossche commented:
@martindurant this will need a merge of master now

Commit:
Merge branch 'master' into feature/add-fsspec-support

martindurant commented:
@jorisvandenbossche , please check carefully; I am guessing at what the new API should be like. Are you certain you want to depend on latest pyarrow here?

jorisvandenbossche commented:
How are we relying on the latest pyarrow? (I might have missed something in the other PR then)

Commit:
account for passed filesystem again

martindurant commented:
Isn't that why the call changed to read_table? After pulling from master, pandas no longer builds for me, which is painful, I supposed some dep changed.
In any case, this will now need massage.

jorisvandenbossche commented:
> Isn't that why the call changed to read_table? 

No, we have been using `read_table` for a long time, it was only recently that a PR changed that to use `ParquetDataset`, but since that broke some use case (like file-like objects), the PR I just merged changed that back to `read_table` (since `read_table` covers all use cases of ParquetDataset, but not the other way around)

Commit:
specify should_close

Commit:
lint

Commit:
Except http passed to fsspec in parquet

(was not previously possible)

Commit:
lint

TomAugspurger commented:
I think one of the test failures in https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=37143&view=logs&j=bef1c175-2c1b-51ae-044a-2437c76fc339&t=770e7bb1-09f5-5ebf-b63b-578d2906aac9 is real.

```
    @tm.network
    @td.skip_if_no("pyarrow")
    def test_parquet_read_from_url(self, df_compat):
        url = (
            "https://raw.githubusercontent.com/pandas-dev/pandas/"
            "master/pandas/tests/io/data/parquet/simple.parquet"
        )
>       df = pd.read_parquet(url)

pandas/tests/io/test_parquet.py:584: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/io/parquet.py:316: in read_parquet
    return impl.read(path, columns=columns, **kwargs)
pandas/io/parquet.py:131: in read
    import_optional_dependency("fsspec")
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
```

Should we not be ussing fsspec when reading from a URL?

martindurant commented:
> Should we not be using fsspec when reading from a URL?

In order not to change current behaviour, http(s) is excepted in get buffer, because it was not handled by fsspec previously.

Commit:
Merge branch 'master' into feature/add-fsspec-support

TomAugspurger commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -297,10 +314,117 @@ Other enhancements
 .. ---------------------------------------------------------------------------
 
-.. _whatsnew_110.api:
+Increased minimum versions for dependencies
```
Looks like a git merge issue.

TomAugspurger commented:
@jorisvandenbossche are you good with this now that https://github.com/pandas-dev/pandas/pull/34500 is in (once the merge issue in the whatsnew is sorted out)?

Commit:
redo whatsnew

jorisvandenbossche commented on pandas/io/parquet.py:
```diff
@@ -115,17 +122,21 @@ def write(
-            self.api.parquet.write_table(
-                table, file_obj_or_path, compression=compression, **kwargs
-            )
+            self.api.parquet.write_table(table, path, compression=compression, **kwargs)
```
It might have been discussed before in this PR, but why this change from `file_obj_or_path` to `path` ? Because with this change, `file_obj_or_path` is basically never used?

martindurant replied:
You know what, this can be made simpler...

The reason is, pyarrow can take a filesystem+path and do the right thing, and for the partitioning case, this is the only way. I think we can deal without opening the file in the pandas code at all. I'll push that suggestion, and add comments for what the blocks do to be clearer.

jorisvandenbossche replied:
The new code above certainly looks fine, I am only wondering if there were things that `get_filepath_or_buffer` was doing before we might now be missing.

It gets a buffer for urls, but I don't think this actually works for *writing*. 
In addition, it also seems to do a `os.path.expanduser` on strings. I suppose this is not tested (and not sure many people would rely on that), but that might lead to a regression.

martindurant replied:
You want to add `expanduser` to the function? I wouldn't bother going through `get_filepath_or_buffer` just for this, and as you say, it isn't tested or documented anywhere. In fact, test may be a little tricky for the various OSs, I'm not sure you can guarantee that `"~/afile"`-> `"${HOME}/afile"`.

TomAugspurger replied:
Confirmed that it works on master

```python
In [2]: df = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})

In [3]: df.to_parquet("~/test-4.parquet")
```

and raises on this branch.

```python
In [3]: df.to_parquet("~/test-5.parquet")
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-3-c7dcbbaae894> in <module>
----> 1 df.to_parquet("~/test-5.parquet")

~/sandbox/pandas/pandas/util/_decorators.py in wrapper(*args, **kwargs)
    197                 else:
    198                     kwargs[new_arg_name] = new_arg_value
--> 199             return func(*args, **kwargs)
    200
    201         return cast(F, wrapper)

~/sandbox/pandas/pandas/core/frame.py in to_parquet(self, path, engine, compression, index, partition_cols, **kwargs)
   2323             index=index,
   2324             partition_cols=partition_cols,
-> 2325             **kwargs,
   2326         )
   2327

~/sandbox/pandas/pandas/io/parquet.py in to_parquet(df, path, engine, compression, index, partition_cols, **kwargs)
    266         index=index,
    267         partition_cols=partition_cols,
--> 268         **kwargs,
    269     )
    270

~/sandbox/pandas/pandas/io/parquet.py in write(self, df, path, compression, index, partition_cols, **kwargs)
    118         else:
    119             # write to single output file
--> 120             self.api.parquet.write_table(table, path, compression=compression, **kwargs)
    121
    122     def read(self, path, columns=None, **kwargs):

~/Envs/pandas-dev/lib/python3.7/site-packages/pyarrow/parquet.py in write_table(table, where, row_group_size, version, use_dictionary, compression, write_statistics, use_deprecated_int96_timestamps, coerce_timestamps, allow_truncated_timestamps, data_page_size, flavor, filesystem, compression_level, **kwargs)
   1341                 use_deprecated_int96_timestamps=use_int96,
   1342                 compression_level=compression_level,
-> 1343                 **kwargs) as writer:
   1344             writer.write_table(table, row_group_size=row_group_size)
   1345     except Exception:

~/Envs/pandas-dev/lib/python3.7/site-packages/pyarrow/parquet.py in __init__(self, where, schema, filesystem, flavor, version, use_dictionary, compression, write_statistics, use_deprecated_int96_timestamps, compression_level, **options)
    434         filesystem, path = resolve_filesystem_and_path(where, filesystem)
    435         if filesystem is not None:
--> 436             sink = self.file_handle = filesystem.open(path, 'wb')
    437         else:
    438             sink = where

~/Envs/pandas-dev/lib/python3.7/site-packages/pyarrow/filesystem.py in open(self, path, mode)
    242         """
    243         path = _stringify_path(path)
--> 244         return open(path, mode=mode)
    245
    246     @property

FileNotFoundError: [Errno 2] No such file or directory: '~/test-5.parquet'
```

pandas consistently expands `~` to the homedir, so we'll need to ensure we don't regress on this.

martindurant replied:
but how do we test this reliably?

TomAugspurger replied:
We should have similar tests for csv.

jorisvandenbossche replied:
We test the common utility: https://github.com/pandas-dev/pandas/blob/80ba4c4294a1ce1d95aeeb8e5de86d33ebb6edf4/pandas/tests/io/test_common.py#L55-L68

and then probaby suppose that all IO methods use those common utilities.

martindurant replied:
A little weak...
I'll put in something. By the way, I expect that expanding the home dir didn't work before either when partitioning.

jorisvandenbossche commented on pandas/io/parquet.py:
```diff
@@ -122,9 +125,17 @@ def write(
-        parquet_ds = self.api.parquet.ParquetDataset(
-            path, filesystem=get_fs_for_path(path), **kwargs
-        )
+        if is_fsspec_url(path) and "filesystem" not in kwargs:
```
Can you check this comment? 
(I will try if I can write a test that would catch it)

jorisvandenbossche replied:
ping for this one

jorisvandenbossche replied:
(I am also fine with doing this as a follow-up myself)

martindurant replied:
I thought you meant that you intended to handle it; and yes please, you are in the best place to check the finer details of the calls to pyarrow.

Commit:
simplify parquet write

Commit:
Retry S3 file probe with timeout, in test_to_s3

TomAugspurger commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1665,13 +1665,21 @@ def test_json_multiindex(self, dataframe, expected):
-            obj.key for obj in s3_resource.Bucket("pandas-test").objects.all()
-        )
+        timeout = 5
+        while True:
```
Hmm, this is concerning. Do you know what's causing it? I would think that everything is synchronous.

martindurant replied:
I don't know why... S3 is not supposed to be immediately consistent, maybe the botocore caches or moto doesn't update it's index immediately, or something like that.

martindurant commented:
(travis is passing, but has not updated in the list above, at least for me)

Commit:
expand user in non-fsspec paths for parquet; add test for this

Commit:
reorder imports!

martindurant commented:
@jorisvandenbossche , anything else?

jorisvandenbossche commented on pandas/io/parquet.py:
```diff
@@ -115,17 +118,21 @@ def write(
+        else:
+            fs = kwargs.pop("filesystem", None)
+            should_close = False
+            path = _expand_user(path)
```
No blocking comment (could also be as a follow-up), but for my understanding (and could maybe use a code comment): what was the reason again that this whole `is_fsspec_url(path): .. else: ..` block cannot be handled by `get_filepath_or_buffer` ? 
(which would eg handle the `_expand_user` as well)

martindurant replied:
That we are not getting an open file, we are passing a path and filesystem to arrow for it to handle. This is important for the partitioning case, but works for the simple case too. Wasn't it your suggestion?

jorisvandenbossche replied:
> Wasn't it your suggestion?

I don't think so, since this was already there from the beginning of the PR before my first review ;) Not that it matters much.

Anyway, to answer my own question (I think): on a second look, this is simply because `get_filepath_or_buffer` doesn't return a filesystem (and which is indeed needed here, as you indicated). I was just confused about that for a moment.

jorisvandenbossche commented:
Thanks!

martindurant commented:
Phew :)

TomAugspurger commented:
Thanks!
